### PR TITLE
Get things working on a basic level for 0.5 and 0.6

### PR DESCRIPTION
**Pull request by kshyatt**
Might need to do some more here...

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -15,1923 +15,1687 @@ end
...
-    @test_approx_eq(A,h_A)
-end
-test_her2!(Complex64)
-test_her2!(Complex128)
-
-##############
-# test gemm! #
-##############
-
-function test_gemm!(elty)
```
Comment (SimonDanisch) [inline test/runtests.jl:963]: Are you just removing these tests?

File: REQUIRE [hunk 2]
```diff
@@ -1,3 +1,3 @@
 Compat
 CUDArt
-julia 0.4-
+julia 0.6-
```
Comment (maleadt) [inline REQUIRE]: What keeps it from being 0.5 compatible?

Comment (kshyatt) [reply REQUIRE]: The fact that I didn't have a chance to test it on 0.5. If it passes there let's relax the restriction.

File: test/runtests.jl (hunk 1 above)
Comment (kshyatt) [reply test/runtests.jl:963]: I don't think so, they should have survived as a `@testset`. Did they not?

File: test/runtests.jl [hunk 3]
```diff
@@ -15,1923 +15,1687 @@ end
...
+        A = (alpha*x)*y' + y*(alpha*x)' + A
+        # move to host and compare upper triangles
+        h_A = to_host(d_A)
+        A = triu(A)
+        h_A = triu(h_A)
+        @test A ≈ h_A
+    end
+end
+
+@testset "gemm!" begin
```
Comment (kshyatt) [inline test/runtests.jl:880]: @SimonDanisch here are the `gemm!` tests

File: test/runtests.jl (hunk 1 above)
Comment (SimonDanisch) [reply test/runtests.jl:963]: ah excellent, I just searched for gemm! in the diff and somehow nothing turned up ;)

**Comment by kshyatt** [comment]
The tests pass locally for me. The problem is with CUDArt.jl.

File: test/runtests.jl [hunk 4]
```diff
@@ -15,1923 +15,1699 @@ end
...
+        @test ndims(A) == 1
+        @test ndims(B) == 1
+        @test length(A) == length(B)
+        n1 = length(A)
+        d_A = CudaArray(A)
+        d_B = CudaArray(B)
+        cuda_dot1 = CUBLAS.dotu(n1,d_A,1,d_B,1)
+        cuda_dot2 = CUBLAS.dotu(d_A,d_B)
+        host_dot = A.'*B
+        if VERSION < v"0.6.0"
```
Comment (maleadt) [inline test/runtests.jl]: `< v"0.6.0-pre"` ?

File: test/runtests.jl [hunk 5]
```diff
@@ -15,1923 +15,1699 @@ end
...
+        @test ndims(A) == 1
+        @test ndims(B) == 1
+        @test length(A) == length(B)
+        n1 = length(A)
+        d_A = CudaArray(A)
+        d_B = CudaArray(B)
+        cuda_dot1 = CUBLAS.dotc(n1,d_A,1,d_B,1)
+        cuda_dot2 = CUBLAS.dotc(d_A,d_B)
+        host_dot = A'*B
+        if VERSION < v"0.6.0"
```
Comment (maleadt) [inline test/runtests.jl]: `< v"0.6.0-pre"` ?

File: test/runtests.jl [hunk 6]
```diff
@@ -15,1923 +15,1699 @@ end
...
+        A = [rand(elty,m,k) for i in 1:10]
+        B = [rand(elty,k,n) for i in 1:10]
+        C = [rand(elty,m,n) for i in 1:10]
+        # move to device
+        d_A = CudaArray{elty, 2}[]
+        d_B = CudaArray{elty, 2}[]
+        d_C = CudaArray{elty, 2}[]
+        for i in 1:length(A)
+            push!(d_A,CudaArray(A[i]))
+            push!(d_B,Cud
```
Comment (maleadt) [inline test/runtests.jl]: These two comparisons fail on my system. `rtol=1e-2` makes it pass

File: test/runtests.jl [hunk 7]
```diff
@@ -15,1923 +15,1699 @@ end
...
+
+@testset "asum" begin
+    @testset for elty in [Float32, Float64, Complex64, Complex128]
+        A = rand(elty, m)
+        @test ndims(A) == 1
+        n1 = length(A)
+        d_A = CudaArray(A)
+        cuda_asum1 = CUBLAS.asum(n1,d_A,1)
+        cuda_asum2 = CUBLAS.asum(d_A)
+        host_asum = sum(abs(real(A)) + abs(imag(A)))
```
Comment (maleadt) [inline test/runtests.jl]: `abs.` ?

File: test/runtests.jl [hunk 8]
```diff
@@ -15,1923 +15,1699 @@ end
...
+    test_axpy!_4(2.0f0+im*2.0f0,rand(Complex64,m),rand(Complex64,m))
+    test_axpy!_4(2.0+im*2.0,rand(Complex128,m),rand(Complex128,m))=#
+end
+
+@testset "iamax and iamin" begin
+    @testset for elty in [Float32, Float64, Complex64, Complex128]
+        A = rand(elty, m)
+        n1 = length(A)
+        d_A = CudaArray(A)
+        Aabs = blasabs(A)
```
Comment (maleadt) [inline test/runtests.jl:235]: `blasabs.` ?

File: test/runtests.jl (hunk 6 above)
Comment (maleadt) [inline test/runtests.jl]: Same with these two, passes with `rtol=1e-2`

**Comment by maleadt** [comment]
Huh, GitHub completely  botched the line numbers on my review comments. The comparison in need of an `rtol` are the `C[:L] ≈ dL` and `C[:U] ≈ dU` checks.

File: test/runtests.jl (hunk 8 above)
Comment (kshyatt) [reply test/runtests.jl:235]: That's a user-defined function at the top of the file, but it needs an `abs.` inside :)

**Comment by kshyatt** [comment]
OK to merge?

File: test/runtests.jl [hunk 9]
```diff
@@ -8,1930 +8,1706 @@ n = 35
...
+        n1 = length(A)
+        d_A = CudaArray(A)
+        d_B = CudaArray(B)
+        cuda_dot1 = CUBLAS.dot(n1,d_A,1,d_B,1)
+        cuda_dot2 = CUBLAS.dot(d_A,d_B)
+        host_dot = dot(A,B)
+        @test host_dot ≈ cuda_dot1
+        @test host_dot ≈ cuda_dot2
+
+        #d_A = CudaArray(A)
```
Comment (vchuravy) [inline test/runtests.jl:76]: Why are these tests commented out?

File: test/runtests.jl [hunk 10]
```diff
@@ -8,1930 +8,1706 @@ n = 35
...
+        B1 = to_host(d_B1)
+        host_axpy = alpha*A + B
+        @test host_axpy ≈ B1
+    end
+    test_axpy!_2(2.0f0,rand(Float32,m),rand(Float32,m))
+    test_axpy!_2(2.0,rand(Float64,m),rand(Float64,m))
+    test_axpy!_2(2.0f0+im*2.0f0,rand(Complex64,m),rand(Complex64,m))
+    test_axpy!_2(2.0+im*2.0,rand(Complex128,m),rand(Complex128,m))
+
+    #=function test_axpy!_3(alpha,A,B)
```
Comment (vchuravy) [inline test/runtests.jl:196]: Same here. Would be nice to have a comment explaining why they need to be commented out

File: src/blas.jl [hunk 11]
```diff
@@ -139,7 +139,7 @@ for (jname, fname, elty) in ((:dot,:cublasDdot_v2,:Float64),
                         incx::Integer,
                         DY::Union{CudaPtr{$elty},CudaArray{$elty}},
                         incy::Integer)
-            result = Array($elty,1)
+            result = Array{$elty}(1)
```
Comment (vchuravy) [inline src/blas.jl:142]: These could probably be `Ref`s

**Review by vchuravy — APPROVED** [review]
Looks good to me. The comments I made are not critical and should not hold this up.

**Review by maleadt — APPROVED** [review]

File: test/runtests.jl (hunk 9 above)
Comment (kshyatt) [reply test/runtests.jl:76]: Because the method (`dot` with `Range`) is broken in `src`

**Comment by SimonDanisch** [comment]
Yay!i should check out if this works with gpuarrays now! :)

**Comment by SimonDanisch** [comment]
Thanks everyone!!
```Julia
julia> Pkg.test("CUBLAS")
INFO: Testing CUBLAS
WARNING: could not import Base.scale into CUBLAS
Test Summary: | Pass  Total
blascopy!     |    8      8
Test Summary: | Pass  Total
scal!         |   24     24
Test Summary: | Pass  Total
dot           |   10     10
Test Summary: | Pass  Total
dotu          |   12     12
Test Summary: | Pass  Total
dotc          |   12     12
Test Summary: | Pass  Total
nrm2          |   16     16
Test Summary: | Pass  Total
asum          |   12     12
Test Summary: | Pass  Total
axpy!         |   16     16
Test Summary:   | Pass  Total
iamax and iamin |    8      8
Test Summary: | Pass  Total
gemv!         |   24     24
Test Summary: | Pass  Total
gemv          |   36     36
Test Summary: | Pass  Total
gbmv!         |   12     12
Test Summary: | Pass  Total
gbmv          |    4      4
Test Summary: | Pass  Total
symv!         |    4      4
Test Summary: | Pass  Total
symv          |    4      4
Test Summary: | Pass  Total
hemv!         |    2      2
Test Summary: | Pass  Total
hemv          |    2      2
Test Summary: | Pass  Total
sbmv!         |    4      4
Test Summary: | Pass  Total
sbmv          |    4      4
Test Summary: | Pass  Total
hbmv!         |    4      4
Test Summary: | Pass  Total
hbmv          |    4      4
Test Summary: | Pass  Total
tbmv!         |    8      8
Test Summary: | Pass  Total
tbmv          |    8      8
Test Summary: | Pass  Total
tbsv!         |    8      8
Test Summary: | Pass  Total
tbsv          |    8      8
Test Summary: | Pass  Total
trmv!         |    4      4
Test Summary: | Pass  Total
trmv          |    4      4
Test Summary: | Pass  Total
trsv!         |    4      4
Test Summary: | Pass  Total
trsv          |    4      4
Test Summary: | Pass  Total
ger!          |    4      4
Test Summary: | Pass  Total
syr!          |    4      4
Test Summary: | Pass  Total
her!          |    2      2
Test Summary: | Pass  Total
her2!         |    2      2
Test Summary: | Pass  Total
gemm!         |    8      8
Test Summary: | Pass  Total
gemm          |    8      8
Test Summary: | Pass  Total
gemm_batched! |   40     40
Test Summary: | Pass  Total
gemm_batched  |   40     40
Test Summary: | Pass  Total
symm!         |    4      4
Test Summary: | Pass  Total
symm          |    4      4
Test Summary: | Pass  Total
syrk!         |    4      4
Test Summary: | Pass  Total
syrk          |    4      4
Test Summary: | Pass  Total
herk!         |    2      2
Test Summary: | Pass  Total
herk          |    2      2
Test Summary: | Pass  Total
syr2k!        |    4      4
Test Summary: | Pass  Total
syr2k         |    4      4
Test Summary: | Pass  Total
her2k!        |    2      2
Test Summary: | Pass  Total
her2k         |    2      2
Test Summary: | Pass  Total
trmm!         |    4      4
Test Summary: | Pass  Total
trmm          |    4      4
Test Summary: | Pass  Total
trsm!         |    4      4
Test Summary: | Pass  Total
trsm          |    4      4
Test Summary: | Pass  Total
trsm_batched! |   40     40
Test Summary: | Pass  Total
trsm_batched  |   40     40
Test Summary: | Pass  Total
hemm!         |    4      4
Test Summary: | Pass  Total
hemm          |    4      4
Test Summary: | Pass  Total
geam!         |   28     28
Test Summary: | Pass  Total
geam          |    4      4
Test Summary:  | Pass  Total
getrf_batched! |  160    160
Test Summary: | Pass  Total
getrf_batched |   80     80
Test Summary: | Pass  Total
getri_batched |  120    120
Test Summary:  | Pass  Total
matinv_batched |   40     40
Test Summary:  | Pass  Total
geqrf_batched! |   40     40
Test Summary: | Pass  Total
geqrf_batched |   40     40
Test Summary: | Pass  Total
gels_batched! |   40     40
Test Summary: | Pass  Total
gels_batched  |   40     40
Test Summary: | Pass  Total
dgmm!         |   12     12
Test Summary: | Pass  Total
dgmm          |    4      4
INFO: CUBLAS tests passed
```
image

**Pull request merged**